File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/monsters/SpeciesDescriptionService.java
package de.limbusdev.guardianmonsters.guardians.monsters;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.utils.ArrayMap;
import com.badlogic.gdx.utils.JsonValue;


/**
 * GuardianDescriptionService
 *
 * @author Georg Eckert 2018
 */

public class SpeciesDescriptionService implements ISpeciesDescriptionService
{
    private ArrayMap<Integer, SpeciesDescription> speciesDB;

    private static SpeciesDescriptionService instance;

    private SpeciesDescriptionService(String jsonSpeciesDescriptions)
    {
        speciesDB = new ArrayMap<>();
        JsonValue rootElement = JSONGuardianParser.parseGuardianList(jsonSpeciesDescriptions);

        for (int i = 0; i < rootElement.size; i++)
        {
            SpeciesDescription info = JSONGuardianParser.parseGuardian(rootElement.get(i));
            speciesDB.put(info.getID(),info);
        }
    }

    public static SpeciesDescriptionService getInstance(String jsonSpecies)
    {
        if(instance == null) {
            instance = new SpeciesDescriptionService(jsonSpecies);
        }
        return instance;
    }

    public static SpeciesDescriptionService getInstanceFromFile(String jsonSpeciesDescriptionFilePath)
    {
        String xml = Gdx.files.internal(jsonSpeciesDescriptionFilePath).readString();
        return getInstance(xml);
    }


    @Override
    public SpeciesDescription getSpeciesDescription(int speciesID)
    {
        return speciesDB.get(speciesID);
    }

    @Override
    public String getCommonNameById(int speciesID, int form)
    {
        return speciesDB.get(speciesID).getNameID(form);
    }


    @Override
    public void destroy()
    {
        instance = null;
    }
}

File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/monsters/SpeciesDescriptionService.kt
package de.limbusdev.guardianmonsters.guardians.monsters

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.utils.ArrayMap
import de.limbusdev.utils.extensions.set


/**
 * GuardianDescriptionService
 *
 * @author Georg Eckert 2019
 */

class SpeciesDescriptionService private constructor (jsonSpeciesDescriptions: String)
    : ISpeciesDescriptionService
{
    private val speciesDB: ArrayMap<Int, SpeciesDescription> = ArrayMap()

    init
    {
        val rootElement = JSONGuardianParser.parseGuardianList(jsonSpeciesDescriptions)

        for (i in 0 until rootElement.size)
        {
            val info = JSONGuardianParser.parseGuardian(rootElement.get(i))
            speciesDB[info.ID] = info
        }
    }

    override fun getSpeciesDescription(speciesID: Int): SpeciesDescription
    {
        return speciesDB.get(speciesID)
    }

    override fun getCommonNameById(speciesID: Int, form: Int): String
    {
        return speciesDB.get(speciesID).getNameID(form)
    }

    override fun destroy() { instance = null }

    companion object
    {
        private var instance: SpeciesDescriptionService? = null

        fun getInstance(jsonSpecies: String): SpeciesDescriptionService
        {
            if (instance == null) { instance = SpeciesDescriptionService(jsonSpecies) }
            return instance!!
        }

        fun getInstanceFromFile(jsonSpeciesDescriptionFilePath: String): SpeciesDescriptionService
        {
            val json = Gdx.files.internal(jsonSpeciesDescriptionFilePath).readString()
            return getInstance(json)
        }
    }
}
